### PR TITLE
HiveHook fix get_pandas_df() failure when it tries to read an empty table

### LINKED ISSUE #17765: get_pandas_df() fails when it tries to read an empty table
**Apache Airflow version**:

2.2.0.dev0

**OS**:

Debian GNU/Linux

**Apache Airflow Provider versions**:

**Deployment**:

docker-compose version 1.29.2

**What happened**:

Currently in hive hooks, when [get_pandas_df()](https://github.com/apache/airflow/blob/main/airflow/providers/apache/hive/hooks/hive.py#L1039) is used to create a dataframe, the first step creates the dataframe, the second step creates the columns.  The step to add columns throws an exception when the table is empty.

```
hh = HiveServer2Hook()
sql = "SELECT * FROM <table> WHERE 1=0"
df = hh.get_pandas_df(sql)

[2021-08-15 21:10:15,282] {{hive.py:449}} INFO - SELECT * FROM <table> WHERE 1=0
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
  File "/venv/lib/python3.7/site-packages/airflow/providers/apache/hive/hooks/hive.py", line 1073, in get_pandas_df
    df.columns = [c[0] for c in res['header']]
  File "/venv/lib/python3.7/site-packages/pandas/core/generic.py", line 5154, in __setattr__
    return object.__setattr__(self, name, value)
  File "pandas/_libs/properties.pyx", line 66, in pandas._libs.properties.AxisProperty.__set__
  File "/venv/lib/python3.7/site-packages/pandas/core/generic.py", line 564, in _set_axis
    self._mgr.set_axis(axis, labels)
  File "/venv/lib/python3.7/site-packages/pandas/core/internals/managers.py", line 227, in set_axis
    f"Length mismatch: Expected axis has {old_len} elements, new "
ValueError: Length mismatch: Expected axis has 0 elements, new values have 1 elements
```

**What you expected to happen**:

When an empty table is read, get_pandas_df() should return an empty dataframe, with the columns listed in the empty dataframe.  Ideally, `len(df.index)` produces `0` and `df.columns` lists the columns.

**How to reproduce it**:

Use your own hive server...
```
from airflow.providers.apache.hive.hooks.hive import HiveServer2Hook
hh = HiveServer2Hook()
sql = "SELECT * FROM <table> WHERE 1=0"
df = hh.get_pandas_df(sql)
```
Where \<table> can be any table.

**Anything else we need to know**:

Here's a potential fix ...
```
    #df = pandas.DataFrame(res['data'], **kwargs)
    #df.columns = [c[0] for c in res['header']]
    df = pandas.DataFrame(res['data'], columns=[c[0] for c in res['header']], **kwargs)

```

A PR is more or less ready but someone else should confirm this is an actual bug.  I can trigger the bug on an actual hive server.  I cannot trigger the bug using [test_hive.py](https://github.com/apache/airflow/blob/main/tests/providers/apache/hive/hooks/test_hive.py).  Also, if a PR Is needed, and a new test is needed, then I may need some help updating test_hive.py.

**Are you willing to submit a PR?**

Yes.

### PR DESCRIPTION
**Pull request by iblaine**
closes: #17765

bug fix for [get_pandas_df()](https://github.com/apache/airflow/blob/main/airflow/providers/apache/hive/hooks/hive.py#L1039) to avoid an exception when reading an empty table.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

**Comment by iblaine** [comment]
Do we want to add a new test to [test_hive.py](https://github.com/apache/airflow/blob/main/tests/providers/apache/hive/hooks/test_hive.py)? I'm not convinced one is needed because this bug is subtle.  And I couldn't figure out how to mock an empty table to trigger this bug.

**Comment by uranusjr** [comment]
I wonder if it’d work to do `SELECT * FROM table LIMIT 0`

**Comment by iblaine** [comment]
> I wonder if it’d work to do `SELECT * FROM table LIMIT 0`

Mock seems to return a result set regardless of the SQL.

**Comment by iblaine** [comment]
Setting this [line](https://github.com/apache/airflow/blob/main/tests/test_utils/mock_process.py#L59) to `self.iterable = []` will trigger this bug.

**Comment by iblaine** [comment]
* A new flag `empty_table_flag` has been added to enable tests against an empty hive table
* A new test has been added to validate get_pandas_df against an empty table

File: tests/providers/apache/hive/hooks/test_hive.py [hunk 1]
```diff
@@ -687,6 +687,13 @@ def test_get_pandas_df(self):
         hook.mock_cursor.execute.assert_any_call('set airflow.ctx.dag_owner=airflow')
         hook.mock_cursor.execute.assert_any_call('set airflow.ctx.dag_email=test@airflow.com')
 
+        hook = MockHiveServer2Hook(empty_table_flag = True)
```
Comment (uranusjr) [inline tests/providers/apache/hive/hooks/test_hive.py]: I’d probably do something like `MockHiveServer2Hook(connection_cursor=EmptyMockTableCursor())` instead. `EmptyMockTableCursor` would inherit from a new `BaseMockHiveServer2Hook` class that contains common logic refactored out of `MockHiveServer2Hook` (which would then also inherit from `BaseMockHiveServer2Hook`).

**Review by uranusjr — CHANGES_REQUESTED** [review]

File: tests/providers/apache/hive/hooks/test_hive.py (hunk 1 above)
Comment (iblaine) [reply tests/providers/apache/hive/hooks/test_hive.py]: Let's see how these latest changes work. I removed the `empty_table_flag ` flag & split up the `MockConnectionCursor` class.  It does look better.

File: tests/test_utils/mock_process.py [hunk 2]
```diff
@@ -85,3 +84,15 @@ def __next__(self):
 
     def __iter__(self):
         yield from self.iterable
+
+
+class MockConnectionCursor(BaseMockConnectionCursor):
+    def __init__(self, *args, **kwargs):
+        super().__init__()
+        self.iterable = [(1, 1), (2, 2)]
```
Comment (uranusjr) [inline tests/test_utils/mock_process.py]: Could you remove all the `*args, **kwargs` arguments from the classes? I know they were there before this patch, but they are not needed and only require unnecessary mental overhead to read.

Comment (iblaine) [reply tests/test_utils/mock_process.py]: Good catch.  Unnecessary args, kwargs arguments have been removed from `mock_process.py`.

**Review by uranusjr — APPROVED** [review]
That transaction failure happens from time to time, likely some sort of race condition in the test runner. We haven't identified the cause yet but it's not related to this change. m

**Comment by iblaine** [comment]
We have a few of new failures.  These new failures appear to be unrelated to this PR.  Should I rebase and try this again, or something else?
* [tests/providers/amazon/aws/operators/test_sqs.py](tests/providers/amazon/aws/operators/test_sqs.py) - test_execute_success()
* [tests/www/test_security.py](tests/www/test_security.py) - test_current_user_has_permissions()

**Comment by uranusjr** [comment]
The second one is also failing in main and there’s a PR to fix it, so don’t worry about it. I think the first one is already fixed; maybe try rebasing the PR to latest main?

**Comment by potiuk** [comment]
Only failures that have been fixed recently in main. Merging now to regenerate providers doc.

**Pull request merged**